Translate Tutorial 2 of 2 (#68)
reactjs/pt-BR.reactjs.org: content/tutorial/tutorial.md

}
```

Nos modificamos `this.props` para `props` nas duas vezes que ela aparece.

**[Veja o código completo até aqui](https://codepen.io/gaearon/pen/QvvJOv?editors=0010)**

>Nota
>

> >Quando modificamos Square para ser um componente funcional, também modificamos `onClick={() => this.props.onClick()}` para uma versão mais curta: `onClick={props.onClick}` (note a ausência dos parenteses em *ambos* os lados). Em uma classe, nós utilizamos uma arrow function para acessar o valor correto de `this`, mas para um componente funcional não precisamos nos preocupar com `this`.

Reviewer: ```suggestion
Quando modificamos Square para ser um componente funcional, também modificamos `onClick={() => this.props.onClick()}` para uma versão mais curta: `onClick={props.onClick}` (note a ausência dos parenteses em *ambos* os lados). Em uma classe, nós utilizamos uma arrow function para acessar o valor correto de `this`. Mas, para um componente funcional não precisamos nos preocupar com `this`.
```

Author: Oi @glaucia86 
É pra remover o ">"  da nota mesmo? Esse texto faz parte de uma nota que é recuada, veja no original.

Essa parte aqui tem o polêmico "componente funcional" vs. "componente de função", acho que é outro termo que precisamos bater o martelo para prosseguirmos, né?

Reviewer 2: Acho que precisamos deixar o `>`

### Trocando Turnos  {#trocando-turnos}

> Agora precisamos consertar um defeito óbvio em nosso jogo de Damas: os "O"s não podem ser marcados no tabuleiro.

Reviewer 3: É um Jogo da Velha, certo? 😆 

```suggestion
Agora precisamos consertar um defeito óbvio em nosso Jogo da Velha: os "O"s não podem ser marcados no tabuleiro.
```

Author: Ahahahaha!! Cara, não sei de onde diabos eu tirei Damas, sei lá pq eu fiquei com "checkers" na cabeça. 

Pior que o restante do arquivo tá todo com "Damas" ahahah! Amanhã eu arrumo isso logo tb!

  }
```

> Com esse mudança,"X"s e "O"s podem trocar os turnos. Também vamos modificar o texto de "status" na função `render` do Board para que ela passe a exibir quem jogará o próximo turno. 

Reviewer 4: ```suggestion
Com esse mudança, "X"s e "0"s podem trocar os turnos. Também vamos modificar o texto de "status" na função `render` do Board para que ela passe a exibir quem jogará o próximo turno. 
```

Author: Concordo com o espaço depois da vírgula, mas o "O" era a letra mesmo no original e não um "0" (número), algum motivo específico para mudar?

Reviewer 4: @halian-vilela Pura verdade! ficou um pouco confuso, pelo fato de que, em várias partes usa-se `0 (zero)`

Author: @thadeu 

Ah! Agora eu vi, eu q mandei mal e troquei alguns por zeros, vou passar um pente fino e corrigir todo, mas vou manter o O (letra) pois é o que eles utilizam no código de exemplo.

Blz?

>       // the rest has not changed

Reviewer 4: ```suggestion
      // o restante continua igual
```

>       // the rest has not changed

Reviewer 4: ```suggestion
      // o restante não tem alterações
```

>       // the rest has not changed

Reviewer 4: ```suggestion
      // o restante não tem alterações
```

}
```

**[Veja o código completo nessa etapa](https://codepen.io/gaearon/pen/KmmrBy?editors=0010)**

### Declarando um Vencedor {#declarando-um-vencedor}

> Agora que temos que mostrar quem jogará o próximo turno, devemos também mostrar quando o jogo foi vencido e que não há mais turnos a serem jogados. Podemos determinar um vencedor adicionando essa função auxiliar ao final do arquivo. 

Reviewer 3: ```suggestion
Agora que mostramos quem jogará o próximo turno, também deveríamos mostrar quando o jogo foi vencido e que não há mais turnos a serem jogados. Podemos determinar um vencedor adicionando essa função auxiliar ao final do arquivo:
```

>       // the rest has not changed

Reviewer 4: ```suggestion
      // o restante não tem alterações
```

      // the rest has not changed
```

> Agora podemos modificar a função `handleClick` do Board para retornar antes, ignorando o click, caso alguém tanha vencido o jogo 

Reviewer 3: ```suggestion
Agora podemos modificar a função `handleClick` do Board para retornar antes, ignorando o click, caso alguém tenha vencido o jogo ou se o quadrado (square) já esteja ocupado:
```

Há uma nova linha entre "jogo" e "ou", onde deveria ter um espaço.

> ou se o quadrado (square) já esteja ocupado:

Reviewer 3: ```suggestion

```

  }
```

**[Veja o código completo até aqui](https://codepen.io/gaearon/pen/LyyXgK?editors=0010)**

Parabéns! Você agora tem um Jogo da Velha funcionando! E também acaba de aprender o básico de React. Então, *você* é provavelmente o verdadeiro vencedor aqui.

> ## Adicionando a Viagem no Tempo (Time Travel){#adding-time-travel}

Reviewer: @halian-vilela time travel é uma expressão idiomática que significa `voltar no tempo` ou `máquina do tempo`. Teríamos que ver como ficaria o contexto dessas traduções nessa parte da documentação

cc: @cezaraugusto @eduardomoroni @fjoshuajr podem ajudar nessa parte

Author: @glaucia86 

Eu concordo que podemos utilizar o contexto, apesar de ter traduzido eu até proporia deixar "Time Travel" mesmo, sem traduzir; uma vez que é uma feature relativamente conhecida em alguns paradigmas de programação e tem palavras relativamente simples de entender até para quem não sabe muito de inglês.

Só discordo que a expressão signifique **voltar no tempo**, uma vez que a viagem possa ser tanto para o passado quanto para o futuro.

Até mesmo dentro do contexto da feature, a ideia é justamente permitir idas e vindas nos estados da aplicação.

Já "máquina do tempo" eu creio que seria melhor representada por "Time Machine" mesmo... vide o livro clássico do H.G. Wells.

Vamos aguardar a opinião dos demais. 😄

Reviewer 2: Eu acho que poderiamos fazer a tradução com a expressão em ingles entre parenteses. Do jeito que está.
Eu particularmente quase nunca penso em `máquina do tempo` quando leio time travel. Colocar o termo "voltar no tempo" acho que não encaixaria bem neste contexto.

  }
```

> **[Veja o código completo nessa etapa](https://codepen.io/gaearon/pen/LyyXgK?editors=0010)**

Reviewer 3: ```suggestion
**[Veja o código completo até aqui](https://codepen.io/gaearon/pen/LyyXgK?editors=0010)**
```

> Parabéns! Você agora tem um jogo de Damas funcionando! E também acaba de aprender o básico de React. Então, *você* é provavelmente o verdadeiro vencedor aqui.

Reviewer 3: ```suggestion
Parabéns! Você agora tem um Jogo da Velha funcionando! E também acaba de aprender o básico de React. Então, *você* é provavelmente o verdadeiro vencedor aqui.
```

> ## Adicionando a viagem no tempo (Time Travel){#adicionando-a-viagem-no-tempo}

Reviewer 3: ```suggestion
## Adicionando a Viagem no Tempo (Time Travel){#adding-time-travel}
```

Como um último exercício, vamos tornar possível fazer uma "volta no tempo" até os movimentos anteriores que aconteceram no jogo.

> ### Armazenando um histórico de movimentos{#armazenando-um-histórico-de-movimentos}

Reviewer 3: ```suggestion
### Armazenando um Histórico de Jogadas {#storing-a-history-of-moves}
```

> Se nós tivéssemos modificdo o array `squares`, a implementação do time travel seria muito difícil.

Reviewer 3: ```suggestion
Se nós tivéssemos modificado o array `squares`, a implementação da volta no tempo seria muito difícil.
```

> If we mutated the `squares` array, implementing time travel would be very difficult.

Reviewer 3: ```suggestion

```

> No entanto, nós utilizamo `slice()` para criar uma nova cópia do array `squares` após cada movimento, e [tratamos ele como imutável](#porque-imutabilidade-e-importante). Isso permitirá que armazenemos cada versão anterior do array `squares` e possamos navegar entre os turnos que já tenham acontecido. 

Reviewer 3: ```suggestion
No entanto, nós utilizamos `slice()` para criar uma nova cópia do array `squares` após cada movimento e [tratamos ele como imutável](#why-immutability-is-important). Isso nos permitirá o armazenamento de cada versão anterior do array `squares` e que possamos navegar entre os turnos que já tenham acontecido. 
```

  }
```

>Nota
>
>Ao contrário do método de arrays `push()`, que você talvez possa estar mais familiarizado, o método `concat()` não modifica o array original, por isso preferimos utilizá-lo. 

Nesse ponto, o componente Board necessita apenas dos métodos `renderSquare` e `render`. O state do jogo e o método `handleClick` devem estar no componente Game.

**[Veja o código completo até aqui](https://codepen.io/gaearon/pen/EmmOqJ?editors=0010)**

### Mostrando as Jogadas Anteriores{#showing-the-past-moves}

Uma vez que estamos gravando o histórico do Jogo da Velha, agora podemos mostrá-lo para o jogador como uma lista de jogadas anteriores.

Aprendemos anterioremente que os elementos React são objetos JavaScript de primeira classe; podemos passá-los livremente por nossas aplicações. Para renderizar múltiplos itens em React, podemos utilizar um array de elementos React.

> Em JavaScript, arrays possuem um [método `map()](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/map) que é normalmente utilizado para mapear uma fonte de dados para outra fonte de dados, por exemplo:

Reviewer 5: faltou um acento grave (back-tick) pra fechar o bloco de código aberto pro `map()`

Reviewer 6: Ajustado no seguinte PR: https://github.com/reactjs/pt-BR.reactjs.org/pull/229 aguardando revisão...

obrigado @micalevisk

  }
```

Por fim, modificaremos o método `render` do componente Game para deixar de renderizar sempre o último movimento e passar a renderizar apenas o movimento selecionado atualmente, de acordo com `stepNumber`:

```javascript{3}
  render() {
    const history = this.state.history;
    const current = history[this.state.stepNumber];
    const winner = calculateWinner(current.squares);

    // o resto não foi modificado
```

Se clicarmos em qualquer passo no histórico do jogo, o tabuleiro do Jogo da Velha deve atualizar imediatamente para mostrar como ficou depois que aquele passo ocorreu.

**[Veja o código completo nessa etapa](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**

### Recapitulando {#wrapping-up}

Parabéns! Você criou um jogo que:

* Te permite jogar o Jogo da Velha,
* Indica quando um dos jogadores ganhou o jogo,

> * Armazena um histórico do jogo à medida que ele progride,

Reviewer 4: ```suggestion
* Armazena um histórico do jogo à medida que ele avança,
```

* Permite aos jogadores revisarem o histórico do jogo e verem versões anteriores do tabuleiro.

Belo trabalho! Esperamos que agora você esteja sentindo como se tivesse uma boa noção de como React funciona.

Dê uma olhada on resultado final aqui: **[Resultado Final](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**.

Se você tiver algum tempo extra e quiser praticar suas habilidades no React, aqui estão algumas ideias de melhorias que você poderia adicionar a seu Jogo da Velha, listadas em ordem crescente de dificuldade.

1. Mostrar a localização de cada movimento no formato (col,row), para cada movimento no histórico.
2. Estilizar com negrito o item da lista de movimentos que está selecionado no momento.

> 3. Rescrever o componente Board para utilizar 2 loops para fazer os quadrados, em vez de deixá-los hardcoded.

Reviewer 4: ```suggestion
3. Reescrever o componente Board para utilizar 2 loops para fazer os quadrados, em vez de deixá-los hardcoded.
```

4. Adicionar um botão de toggle que lhe permita ordenar os movimentos em ordem ascendente ou descendente.
5. Quando alguém ganhar, destaque os 3 quadrados que causaram a vitória.
6. Quando ninguém ganhar, exiba uma mensagem informando que o resultado foi um empate.

> Ao longo dessa tutorial, abordamos conceitos de React incluindo elementos, componentes, props e state. Para uma explicação mais detalhada de cada um desses tópicos, confira [o restante da documentação](/docs/hello-world.html). Para aprender mais sobre definição de componentes, confira a [API de Referência do `React.Component](/docs/react-component.html).

Reviewer 4: ```suggestion
Ao longo dessa tutorial, abordamos conceitos de React incluindo elementos, componentes, props e state. Para uma explicação mais detalhada de cada um desses tópicos, confira [o restante da documentação](/docs/hello-world.html). Para aprender mais sobre definição de componentes, confira a [API de Referência do `React.Component`](/docs/react-component.html).
```

Reviewer 4: @halian-vilela faltou fechar o "`" Component

> Ao longo dessa tutorial, abordamos conceitos de React incluindo elementos, componentes, props e state. Para uma explicação mais detalhada de cada uma desses tópicos, confira o restante da documentação [o restante da documentação](/docs/hello-world.html). Para aprender mais sobre definição de componentes, confira a [referência da API `React.Component](/docs/react-component.html).

Reviewer 4: ```suggestion
Ao longo dessa tutorial, abordamos conceitos de React incluindo elementos, componentes, props e state. Para uma explicação mais detalhada de cada um desses tópicos, confira [o restante da documentação](/docs/hello-world.html). Para aprender mais sobre definição de componentes, confira a [API de Referência do `React.Component](/docs/react-component.html).
```
